### DISCUSSION TITLE
How can I convert TDCS simulation results to MNI space with a custom mesh model?

### DISCUSSION THREAD

**Discussion by Raichuu41** [simnibs/simnibs#230] [9. Results and analysis] [answered]
Hello! I am trying to convert my TDCS simulation results to the MNI space, so that I can have a visually appealing results plot about for example the magnJ field data along an axis. How do I go on about it? I checked the documentation but I don't understand which coordinates I have to pass by?

Assuming that I have my simulations results stored in `simulation_TDCS_1_scalar.msh`, this is what I do:
```python
import simnibs

result_mesh: simnibs.Msh = simnibs.mesh_tools.read_msh("simulation_TDCS_1_scalar.msh")
magn_j: simnibs.ElementData = result_mesh.elmdata[-1]  # last entry is with field name magnJ
mni_coordinates = simnibs.subject2mni_coords(coordinates=magn_j.mesh.nodes.node_coord, m2m_folder="m2m_simulation")
```
Am I passing the right coordinates here? And if yes, what am I still missing to do the transformation? If I run this code, I get multiple warnings that these files could not be found:
1. `m2m_simulation\T1.nii.gz`
2. `m2m_simulation\toMNI\MNI2Conform_nonl.nii.gz`
3. `m2m_simulation\toMNI\Conform2MNI_nonl.nii.gz`

Where do I get these files from? If I run my TDCS simulation with a customly created mesh box, I have exactly just one file in my m2m folder, that is `simulation.msh` (my m2m folder is called `m2m_simulation`). Is there something else I have to create via Python before I can convert the coordinates?
If it matters or helps to understand the initial setup, I am not using SimNIBS for a head model but for a box-shaped model with a specific setup of electrodes placed, here's an image:
![image](https://github.com/simnibs/simnibs/assets/25647046/7b57300f-a4dc-4b5f-82e9-810143bffbed)

On a side note: If there is an easier way to get more details of the simulation results for plotting purposes, feel free to let me know. I don't have to use the MNI space, I just thought that this is what I need for plotting results.

**Comment by bjornsig** [comment]
When creating a head model, charm also creates the files for transformation to MNI space. These are the warnings you get. Unfortunately when you have a custom mesh you don't have the transformations. You don't need to use MNI space to plot results.

**Reply by Raichuu41** [reply]
Thank you for the explanation. How do I plot the results instead? Or rather, where can I read the simulation results to do further calculations? I know that it has to be a combination of the `<...>_scalar.msh` and `<...>.geo` file but what's the exact structure there? Or is there an easier way to get the results? 
I need to do line integrals on the electric field values from point A (coordinate of one electrode) to B (coordinate of another electrode). But I don't know how I can access these values.

**Reply by TorgeW** [reply, marked as answer]
If you want to calculate potential differences you should use the potentials aka line integrals. 
https://simnibs.github.io/simnibs/build/html/documentation/sim_struct/session.html

mesh.field['v'].data is how you can extract the values for every element from the <...>_scalar.msh in python with SimNIBS.

**Reply by Raichuu41** [reply]
Thank you! I wasn't aware of the fields "v". I realised a mistake in the other discussion that I cannot just take the magnE values directly but that I have to do a combination of the E field for the 3D vectors and the magnE values, then calculate vector based of the two intersection points on the triangle element and do the line integral with the magnE value and that intersection vector of each triangle.
If I don't have some thoughts processing error right now, that should change the way I can get the potential differences.

Edit: I am not aware where I am supposed to use `mesh.field["v"].data`? I thought you meant the attached mesh to the SESSION object, so in that sense `SESSION.mesh.field`. But there it is an empty dictionary after `run_simnibs()`.
Anyway I retrieve the data through gmsh, if anyone will need this for the future (simplified):
```python
gmsh.initialize()
gmsh.open(<path to scalar.msh>)
view_tags: list[int] = gmsh.view.getTags()
# create a dict with key as tag name, value as tag number
tag_pairs= {gmsh.view.option.getString(tag, "Name"): tag for tag in view_tags}
# load the data for the view tag (postprocessing views in gmsh)
model_data = gmsh.view.getModelData(tag_pairs["v"], 0)
element_tags = model_data[1]
values = model_data[2]  # the v values
```

**Reply by TorgeW** [reply]
you would load <path to scalar.msh> after the simulation with mesh_io and than you would have the different fields in there.
